File: common/src/main/java/net/polyv/android/player/common/ui/component/PLVMediaPlayerMoreLayoutAudioModeActionView.java
package net.polyv.android.player.common.ui.component;

import static net.polyv.android.player.sdk.foundation.graphics.ColorsKt.parseColor;
import static net.polyv.android.player.sdk.foundation.lang.PreconditionsKt.requireNotNull;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import net.polyv.android.player.business.scene.common.model.vo.PLVMediaOutputMode;
import net.polyv.android.player.common.R;
import net.polyv.android.player.common.di.PLVMediaPlayerLocalProvider;
import net.polyv.android.player.common.modules.media.viewmodel.PLVMPMediaViewModel;
import net.polyv.android.player.common.modules.media.viewmodel.viewstate.PLVMPMediaInfoViewState;
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.PLVMPMediaControllerViewModel;
import net.polyv.android.player.sdk.foundation.lang.PLVRememberState;
import net.polyv.android.player.sdk.foundation.lang.PLVRememberStateCompareResult;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;

/**
 * @author Hoshiiro
 */
public class PLVMediaPlayerMoreLayoutAudioModeActionView extends FrameLayout implements View.OnClickListener {

    private ImageView audioModeActionIv;
    private TextView audioModeActionTv;

    private int tintColorIconNormal = Color.WHITE;
    private int tintColorIconSelected = parseColor("#3F76FC");
    private int textColorNormal = parseColor("#CCFFFFFF");
    private int textColorSelected = parseColor("#CC3F76FC");

    protected PLVMediaOutputMode currentMediaOutputMode = null;

    public PLVMediaPlayerMoreLayoutAudioModeActionView(@NonNull Context context) {
        super(context);
        init(null);
    }

    public PLVMediaPlayerMoreLayoutAudioModeActionView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public PLVMediaPlayerMoreLayoutAudioModeActionView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);
    }

    private void init(@Nullable AttributeSet attrs) {
        LayoutInflater.from(getContext()).inflate(R.layout.plv_media_player_ui_component_more_audio_mode_action_layout, this);
        audioModeActionIv = findViewById(R.id.plv_media_player_audio_mode_action_iv);
        audioModeActionTv = findViewById(R.id.plv_media_player_audio_mode_action_tv);

        parseAttrs(attrs);

        setOnClickListener(this);
    }

    private void parseAttrs(@Nullable AttributeSet attrs) {
        TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.PLVMediaPlayerMoreLayoutAudioModeActionView);
        tintColorIconNormal = typedArray.getColor(R.styleable.PLVMediaPlayerMoreLayoutAudioModeActionView_plvIconTintNormal, tintColorIconNormal);
        tintColorIconSelected = typedArray.getColor(R.styleable.PLVMediaPlayerMoreLayoutAudioModeActionView_plvIconTintSelected, tintColorIconSelected);
        textColorNormal = typedArray.getColor(R.styleable.PLVMediaPlayerMoreLayoutAudioModeActionView_plvTextColorNormal, textColorNormal);
        textColorSelected = typedArray.getColor(R.styleable.PLVMediaPlayerMoreLayoutAudioModeActionView_plvTextColorSelected, textColorSelected);
        typedArray.recycle();
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (isInEditMode()) return;

        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaViewModel.class)
                .getMediaInfoViewState()
                .observeUntilViewDetached(this, new Function1<PLVMPMediaInfoViewState, Unit>() {
                    @Override
                    public Unit invoke(PLVMPMediaInfoViewState viewState) {
                        currentMediaOutputMode = viewState.getOutputMode();
                        onViewStateChanged();
                        return null;
                    }
                });
    }

    protected void onViewStateChanged() {
        PLVRememberState.rememberStateOf(this, "onChangeColor")
                .compareLastAndSet(currentMediaOutputMode)
                .ifNotEquals(new Function1<PLVRememberStateCompareResult, Unit>() {
                    @Override
                    public Unit invoke(PLVRememberStateCompareResult result) {
                        onChangeColor();
                        return null;
                    }
                });
    }

    protected void onChangeColor() {
        if (currentMediaOutputMode == PLVMediaOutputMode.AUDIO_ONLY) {
            audioModeActionIv.setImageTintList(ColorStateList.valueOf(tintColorIconSelected));
            audioModeActionTv.setTextColor(textColorSelected);
        } else {
            audioModeActionIv.setImageTintList(ColorStateList.valueOf(tintColorIconNormal));
            audioModeActionTv.setTextColor(textColorNormal);
        }
    }

    @Override
    public void onClick(View v) {
        boolean currentIsAudioMode = currentMediaOutputMode == PLVMediaOutputMode.AUDIO_ONLY;
        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaViewModel.class)
                .changeMediaOutputMode(currentIsAudioMode ? PLVMediaOutputMode.AUDIO_VIDEO : PLVMediaOutputMode.AUDIO_ONLY);
        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaControllerViewModel.class)
                .popFloatActionLayout();
    }

}

File: common/src/main/java/net/polyv/android/player/common/ui/component/more/PLVMediaPlayerMoreLayoutAudioModeActionView.kt
package net.polyv.android.player.common.ui.component.more

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import net.polyv.android.player.business.scene.common.model.vo.PLVMediaOutputMode
import net.polyv.android.player.common.R
import net.polyv.android.player.common.di.PLVMediaPlayerLocalProvider
import net.polyv.android.player.common.modules.media.viewmodel.PLVMPMediaViewModel
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.PLVMPMediaControllerViewModel
import net.polyv.android.player.sdk.foundation.graphics.parseColor

/**
 * @author Hoshiiro
 */
class PLVMediaPlayerMoreLayoutAudioModeActionView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr), View.OnClickListener {
    private val audioModeActionIv by lazy { findViewById<ImageView>(R.id.plv_media_player_audio_mode_action_iv) }
    private val audioModeActionTv by lazy { findViewById<TextView>(R.id.plv_media_player_audio_mode_action_tv) }

    private var tintColorIconNormal = Color.WHITE
    private var tintColorIconSelected = parseColor("#3F76FC")
    private var textColorNormal = parseColor("#CCFFFFFF")
    private var textColorSelected = parseColor("#CC3F76FC")

    init {
        LayoutInflater.from(context).inflate(R.layout.plv_media_player_ui_component_more_audio_mode_action_layout, this)
        parseAttrs(attrs)
        setOnClickListener(this)
    }

    private fun parseAttrs(attrs: AttributeSet?) {
        val typedArray = context.obtainStyledAttributes(
            attrs,
            R.styleable.PLVMediaPlayerMoreLayoutAudioModeActionView
        )
        tintColorIconNormal = typedArray.getColor(
            R.styleable.PLVMediaPlayerMoreLayoutAudioModeActionView_plv_icon_tint_normal,
            tintColorIconNormal
        )
        tintColorIconSelected = typedArray.getColor(
            R.styleable.PLVMediaPlayerMoreLayoutAudioModeActionView_plv_icon_tint_selected,
            tintColorIconSelected
        )
        textColorNormal = typedArray.getColor(
            R.styleable.PLVMediaPlayerMoreLayoutAudioModeActionView_plv_text_color_normal,
            textColorNormal
        )
        textColorSelected = typedArray.getColor(
            R.styleable.PLVMediaPlayerMoreLayoutAudioModeActionView_plv_text_color_selected,
            textColorSelected
        )
        typedArray.recycle()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (isInEditMode) return

        PLVMediaPlayerLocalProvider.localDependScope.on(this).current()!!
            .get<PLVMPMediaViewModel>()
            .mediaInfoViewState
            .observe { viewState ->
                val currentMediaOutputMode = viewState.outputMode
                if (currentMediaOutputMode == PLVMediaOutputMode.AUDIO_ONLY) {
                    audioModeActionIv.setImageTintList(ColorStateList.valueOf(tintColorIconSelected))
                    audioModeActionTv.setTextColor(textColorSelected)
                } else {
                    audioModeActionIv.setImageTintList(ColorStateList.valueOf(tintColorIconNormal))
                    audioModeActionTv.setTextColor(textColorNormal)
                }
            }
            .disposeOnDetached(this)
    }

    override fun onClick(v: View?) {
        val dependScope = PLVMediaPlayerLocalProvider.localDependScope.on(this).current()
        val currentMediaOutputMode = dependScope?.get<PLVMPMediaViewModel>()?.mediaInfoViewState?.value?.outputMode
        val currentIsAudioMode = currentMediaOutputMode == PLVMediaOutputMode.AUDIO_ONLY
        dependScope?.get<PLVMPMediaViewModel>()
            ?.changeMediaOutputMode(if (currentIsAudioMode) PLVMediaOutputMode.AUDIO_VIDEO else PLVMediaOutputMode.AUDIO_ONLY)
        dependScope?.get<PLVMPMediaControllerViewModel>()
            ?.popFloatActionLayout()
    }
}
